File: scripts/stmt-proc-py/src/processors/statement_processor_provider.py
from .hdfc_bank_acct_processor import HdfcBankAcctStatementProcessor
from .hdfc_credit_card_processor import HdfcCreditCardStatementProcessor
from store.txn_store import TxnStore


class StatementProcessorProvider:
    """Factory class to provide the appropriate statement processor."""

    @staticmethod
    def get_processor(key, txn_store: TxnStore):
        if key == "hdfc-sa":
            return HdfcBankAcctStatementProcessor(txn_store)
        elif key == "hdfc-cc":
            return HdfcCreditCardStatementProcessor(txn_store)
        else:
            raise ValueError(f"Unsupported statement type: {key}")

File: scripts/stmt-proc-wasm/src/statement_processor_provider.rs
use crate::hdfc_bank_acct_processor::HdfcBankAcctStatementProcessor;
use crate::hdfc_credit_card_processor::HdfcCreditCardStatementProcessor;
use crate::statement_processor::StatementProcessor;
use crate::txn_store::TxnStore;

pub struct StatementProcessorProvider;

impl StatementProcessorProvider {
    /// Factory method to get the appropriate statement processor.
    pub fn get_processor(key: &str, txn_store: TxnStore) -> Result<Box<dyn StatementProcessor>, String> {
        match key {
            "hdfc-sa" => Ok(Box::new(HdfcBankAcctStatementProcessor::new(txn_store))),
            "hdfc-cc" => Ok(Box::new(HdfcCreditCardStatementProcessor::new(txn_store))),
            _ => Err(format!("Unsupported statement type: {}", key)),
        }
    }
}